haskell-api.md

```diff
@@ -0,0 +1,236 @@
+
+For example:
+
+```
+/3/teachers
+[ { id:
+  , ...
+  , school:
+    { id:
+    , ...
+    }
+  }
+]
+```
+
+Here, a School filter would be `school.id`.
+
+```
+/3/students
+[ { id:
+  , ...
+  , schools:
+    [ { id:
+      , ...
+      }
+    ]
+  }
+]
+```
+
+But here, it would be `schools.id`.
+
+If a filter exists that is not for an attribute present in the response, the
+name can be inferred by what it would look like if it were.
+
+### Semantics
+
+Prefer `IN` equality semantics.
+
+- Good: `{attribute}={value1},{value2},{value3},...`
+- Bad: `{attribute}={value}`
+
+This is because:
+
+- It's just as easy to support as direct equality
+- It reduces to the expected `EQUALS` semantics when given one value
+- We think we can get by with only this level "smart" filtering for a while and
+  defer more complexity in this area (e.g. "not", etc).
+
+---
+
+**Aside about URL length limits**: We don't know at what size this will be a
+problem. Standards-wise, there is no limit, but browsers and servers do have
+limits and it's recommended to use 2000 as a reference point [ref].
+
+[ref]: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
+
+However, I will say this:
+
+- Our current `/foo/:id,:id,:id/bar` does bump into issues of length and it is
+  annoying to resolve because it means a new *route* to find a natural way to
+  group the massive list (i.e. all students in a school will mean changing
+  `/students/...` to `schools/:id`)
+- Doing the same thing in a `GET` param will be much nicer to handle when we
+  encounter it because we would probably already support both `students.id` and
+  `schools.id` in the single endpoint and the Frontend can just switch
+
+Therefore, I find the length concerns for `IN`-queries to be less problematic
+than our multi-id path-pieces of today.
+
+---
+
+## Request Shape
+
+- Request bodies can vary by use, but this should be avoided if possible
+
+*TODO*: more to say here?
+
+## Response fields
+
+- Always include `id`s
+- Within reason, choose attributes directly for the work that is motivating the
+  endpoint
+- Include related resources in abbreviated form, Frontend almost always wants
+  names:
+  - `students: [{id:, firstName:, lastName:}]`
+  - `school: {id:, name:}`
+
+## Status Codes
+
+- 201 for creation
+- 204 for no content (i.e. our `Empty` responses)
+- 400s MUST follow our `ValidationError` machinery
+
+## Pagination
+
+- All list-returning routes of modest size should be paginated via `Yesod.Page`
+
+## Modules Structure
+
+- Fully expand all parents to start
+
+  This means we won't break Frontend when the tree expands over time
+
+- Fully name-space all parents and routes (see example)
+
+  **Caveats**:
+
+  - There's no need to `V3`-prefix everything
+  - Things can "start over" when there's the "single below a list" situation,
+    e.g. `TeacherP x $ TeacherR` instead of `TeachersP $ TeachersTeacherP x $
```

Comment:
Probably `TeachersP x $ TeacherR`

Reply:
Not if you're following the "expand every parent to start" guide.

```
-- Good
/teachers TeachersP:
  / TeachersR
  /#TeacherId TeacherP:
    / TeacherR GET

-- Bad
/teachers TeachersP:
  / TeachersR
  /#TeacherId TeacheR GET

-- Because good won't break frontend when things expand:
/teachers TeachersP:
  / TeachersR
  /#TeacherId TeacherP:
    / TeacherR GET
    /whatever TeacherWhateverP:
      / TeacherWhateverR GET
```

Reply:
<sup>it would be nice if route generation worked on routes, not their haskell representation :grimacing:</sup>

Another option is to flatten completely, which i'm quietly a fan of:

```haskell
/teachers TeachersR GET 
/teachers/#TeacherId TeacherR GET
/teachers/#TeacherId/whatever TeacherWhateverR GET
```

Don't feel strongly enough to push for that though. What you've stated here makes sense; thanks for thinking about this temporally!

Reply:
I actually don't mind that. It's nice that it requires so many fewer identifies to build the route in Haskell

- `TeachersR`, vs
- `V2P $ TeachersP TeachersR`

Are there downsides?

I can see it'll lead to some **very** long lines in the routes files, too bad we can't do

```
/some/super/long/route/and/more/things/here/omg/80/is/coming/so/coon \
  TheThingR GET POST DELETE
```

Reply:
There may be some issues when you have ambiguity and need to use `!`, but I haven't thought tooooo hard about it.